go/genkit/chat.go

```diff
@@ -0,0 +1,335 @@
+			return nil, err
+		}
+	}
+
+	if chat.Session == nil {
+		s, err := NewSession(ctx)
+		if err != nil {
+			return nil, err
+		}
+		chat.Session = s
+	}
+
+	if chat.ThreadName == "" {
+		chat.ThreadName = "default"
+	}
+
+	return chat, nil
+}
+
+// Send sends a message to the chat, generating a response from the AI model.
+// It retrieves the chat history from the session store, adds the new message
+// to the history, and sends the entire conversation to the AI model for
+// generating a response. If a system message is set for the chat, it is
+// included in the conversation before the history.
+func (c *Chat) Send(ctx context.Context, msg any) (resp *ai.ModelResponse, err error) {
+	// Load history
+	data, err := c.Session.Store.Get(c.Session.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set session details in context
+	ctx = c.Session.SetContext(ctx)
+
+	var messages []*ai.Message
+	if c.SystemText != "" {
+		// Add system message before history
+		messages = append(messages, ai.NewSystemTextMessage(c.SystemText))
+	}
+
+	// Handle prompt if set
+	if c.Prompt != nil {
+		mr, err := c.Prompt.Render(ctx, c.Input)
+		if err != nil {
+			return nil, err
+		}
+		messages = append(messages, mr.Messages...)
+	}
+
+	// Add history
+	messages = append(messages, data.Threads[c.ThreadName]...)
+
+	// Add new message after history
+	aiMsgs, err := getChatMessages(msg)
+	if err != nil {
+		return nil, err
+	}
+	messages = append(messages, aiMsgs...)
+
+	// Assemble options
+	var generateOptions []ai.GenerateOption
+	generateOptions = append(generateOptions, ai.WithModel(c.Model))
+	generateOptions = append(generateOptions, ai.WithMessages(messages...))
+	generateOptions = append(generateOptions, ai.WithConfig(c.Request.Config))
+	generateOptions = append(generateOptions, ai.WithStreaming(c.Stream))
+	generateOptions = append(generateOptions, ai.WithTools(c.Request.Tools...))
+	generateOptions = append(generateOptions, ai.WithContext(c.Request.Context...))
+
+	if c.Request.Format != "" {
+		generateOptions = append(generateOptions, ai.WithOutputFormat(c.Request.Format))
+	}
+	if c.Request.Schema != nil {
+		generateOptions = append(generateOptions, ai.WithOutputSchema(c.Request.Schema))
+	}
+
+	// Call generate
+	resp, err = ai.Generate(ctx,
+		c.Genkit.reg,
+		generateOptions...,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Update history
+	messages = append(messages, resp.Message)
+	err = c.UpdateMessages(c.ThreadName, messages)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+// UpdateMessages updates the messages for the chat.
+func (c *Chat) UpdateMessages(ThreadName string, messages []*ai.Message) error {
+	c.Request.Messages = messages
+	return c.Session.UpdateMessages(ThreadName, messages)
+}
+
+// WithModel sets the model for the chat.
+func WithModel(model ai.Model) ChatOption {
+	return func(c *Chat) error {
+		if c.Model != nil {
+			return errors.New("cannot set model (WithModel) more than once")
```

Comment:
To be consistent with the rest of our API, can you update all of the error messages to follow the format `genkit.WithModel: cannot set Model more than once`.